dot/metrics/collector.go

```diff
@@ -1,32 +1,131 @@
+// Copyright 2019 ChainSafe Systems (ON) Corp.
+// This file is part of gossamer.
+//
+// The gossamer library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The gossamer library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the gossamer library. If not, see <http://www.gnu.org/licenses/>.
+
 package metrics
 
 import (
+	"context"
+	"runtime"
 	"time"
 
 	ethmetrics "github.com/ethereum/go-ethereum/metrics"
 )
 
-// GaugeCollector abstracts the Update function to collectors
-// just implement the collection
-type GaugeCollector interface {
-	Update() int64
+// GaugeMetrics interface allows the exportation of many gauge metrics
+// the implementer could exports
+type GaugeMetrics interface {
+	CollectGauge() map[string]int64
+}
+
+// Collector struct controls the metrics and executes polling to extract the values
+type Collector struct {
+	ctx    context.Context
+	gauges []GaugeMetrics
+}
+
+// NewCollector creates a new Collector
+func NewCollector(ctx context.Context) *Collector {
+	return &Collector{
+		ctx:    ctx,
+		gauges: make([]GaugeMetrics, 0),
+	}
 }
 
-// CollectGaugeMetrics receives an timeout, label and a gauge collector
-// and acquire the metrics timeout by timeout to a ethereum metrics gauge
-func CollectGaugeMetrics(timeout time.Duration, label string, c GaugeCollector) {
-	t := time.NewTicker(timeout)
+// Start will start one goroutine to collect all the gauges registereds and
+// a separate goroutine to collect process metrics
+func (c *Collector) Start() {
```

Comment:
I like making `Start` functions blocking.  It is the callers responsibility to call `go c.Start()`. You could use a `sync.WaitGroup` to wait for these child goroutines to finish.

Reply:
Done!